### PR TITLE
Added 2 fuzzers

### PR DESCRIPTION
**Pull request by AdamKorcz**
Hello

I saw in the code that a fuzzing was on the to-do list:
https://github.com/tailscale/tailscale/blob/e4f53e9b6f1a4d3d6f00091ddef617989b1ea3e4/net/stun/stun_test.go#L16

I have run these on OSS-fuzz's infrastructure as well and would be happy to set up an integration application so they can run continuously free of charge by OSS-fuzz. 

Tagging @bradfitz since you were mentioned in the comment.

Signed-off-by: AdamKorcz <adam@adalogics.com>

### CODE REVIEW

**Comment by bradfitz** [comment]
Thanks!

Assigning to @josharian to review as our resident fuzzy.

File: fuzzing/disco_fuzzer.go [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+package fuzzing
+
+import "tailscale.com/disco"
+
+func FuzzDiscoParser(data []byte) int {
+	_, err := disco.Parse(data)
+	if err != nil {
+		return 0
+	}
```
Comment (josharian) [inline fuzzing/disco_fuzzer.go:13]: Once we have a valid `disco.Message`, we should fuzz using that as well. E.g. we should check that AppendMessage round trips: Use AppendMessage to encode the Message, then parse it again, and then use reflect.DeepEqual to confirm that the twice-parsed message is identical to the once-parsed message.

File: fuzzing/stun_fuzzer.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+package fuzzing
+
+import "tailscale.com/net/stun"
+
+func FuzzStunParser(data []byte) int {
+	_, _, _, err := stun.ParseResponse(data)
+	if err != nil {
+		return 0
+	}
```
Comment (josharian) [inline fuzzing/stun_fuzzer.go:13]: Try stun.ParseBindingRequest here as well?

(And then rename this just Fuzz, and don't worry about returning 0 vs 1. The signal from those is weak, and go-fuzz's support for running multiple fuzz functions is poor enough that it's nicer to just have a single mixed fuzz function.)

File: fuzzing/disco_fuzzer.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package fuzzing
```
Comment (josharian) [inline fuzzing/disco_fuzzer.go:5]: I'd rather this live in the same package as the code that it is fuzzing. You can protect this file with a go-fuzz build tag. Ditto for the other new file.

**Review by josharian — COMMENTED** [review]
Thanks! Always happy to see new fuzz functions. :)

> I have run these on OSS-fuzz's infrastructure as well and would be happy to set up an integration application so they can run continuously free of charge by OSS-fuzz.

That'd be fabulous, thanks. And hopefully there'll be more here as well, over time. Can we arrange to make it easy for Tailscalers (such as me) to update configs as necessary, receive the reports, and generally be looped in?

If you're in the mood, it'd be great to also get OSS-fuzz running the fuzz functions in https://github.com/inetaf/netaddr. :)

> our resident fuzzy

Hmmm...I'm going to interpret that as a comment on how infrequently I shave. Guilty as charged. :P

**Comment by AdamKorcz** [comment]
Hi @josharian It is a pleasure to meet you - I have obviously seen all your great work on the go-fuzz fuzzer.
Thank you for the comments. I will look closer into them over the next days. 

> If you're in the mood, it'd be great to also get OSS-fuzz running the fuzz functions in https://github.com/inetaf/netaddr. :)

I would be glad to. Before I start on this, could you let me know if this could be considered a "critical" project for any organisation? ie: Is it used in production anywhere? It is a question that the OSS-fuzz maintainers would like answered when submitting a new project.

**Comment by josharian** [comment]
>  could you let me know if this could be considered a "critical" project for any organisation? ie: Is it used in production anywhere?

Yes, it provides the core networking types used (heavily) in production at Tailscale.

**Comment by DentonGentry** [comment]
The test failure in DrainPendingMessages isn't from this PR, it was my doing. If you git pull and rebase it should be resolved.

**Comment by AdamKorcz** [comment]
This is by no means abandoned. I should get the comments addressed in the following week.

**Comment by josharian** [comment]
No rush. :)

**Comment by AdamKorcz** [comment]
@josharian Do my latest updates address your comments?

**Pull request merged**